### PR TITLE
XCP-Portscanner added

### PR DESCRIPTION
**Pull request by Ita1pu**
Added Portscanner for XCP #2372 

currently it works under python3 and python2 but unit tests are just on python3 since they are not using python-can.

There are no impacts on other packages

### CODE REVIEW

**Comment by polybassa** [comment]
Hi @Ita1pu , thanks for your PR. Since XCP is similar to CCP, could you please have a look at my CCP implementation. The main focus of a protocol implementation should be the definition of all relevant packets. Additionally, specific scanners with the usage of Scapy internal functions (`sniff, sr1, etc..`) are nice to have in your implementation. I will point out some problems in your current code, but we should discuss the whole implementation.

File: scapy/contrib/automotive/can_xcp.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            self.__message.identifier = i
+            self.__current_id_var.c.acquire()
+            self.__current_id_var.id = i
+            self.__current_id_var.c.release()
+            self.__socket.send(self.__message)
+            time.sleep(self.__cycle_time)
+        self.__console.println("Done")
+
+
+class XCP_SNIFF_THREAD(threading.Thread):
```
Comment (polybassa) [inline scapy/contrib/automotive/can_xcp.py]: Scapy has a very powerful `sniff` function. Please use this function.

File: scapy/contrib/automotive/can_xcp.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            if (msg and data_array and data_array[0] == 0xFF and
+                    msg.identifier not in self.__known_ids and
+                    msg.length == 8 and
+                    msg.identifier != self.__current_id_var.id):
+                self.__console.println(hex(msg.identifier) +
+                                       " " +
+                                       hex(self.__current_id_var.id))
+                request_id = self.__current_id_var.id
+                response_id = msg.identifier
+                max_dto = (data_array[4] << 8) + data_array[5]
```
Comment (polybassa) [inline scapy/contrib/automotive/can_xcp.py]: The interpretation of data should be done through `Packet` definitions. For example: 
```python
class myPacket(Packet):
    field_des = [
        ... some fields ...
    ]

some_socket.basecls = myPacket
# or
raw_data = some_socket.recv() # or some_socket.sniff(...)
pkt = myPacket(raw_data)
pkt.someField == whatever
```

File: test/contrib/automotive/can_xcp.uts [hunk 3]
```diff
@@ -0,0 +1,316 @@
+% Regression tests for the XCP_CAN
```
Comment (polybassa) [inline test/contrib/automotive/can_xcp.uts]: I refactored all my unit tests which use vcan_sockets, recently. I highly recommend to copy my initialisation code from there to have a identical setup, running on all travis machines.

Init-code: https://github.com/secdev/scapy/blob/ff43e450709605314c4739287e83c588a974a502/test/contrib/automotive/ccp.uts#L1-L96

Cleanup-code: https://github.com/secdev/scapy/blob/ff43e450709605314c4739287e83c588a974a502/test/contrib/automotive/ccp.uts#L1044-L1053

Usage example of sockets: https://github.com/secdev/scapy/blob/ff43e450709605314c4739287e83c588a974a502/test/contrib/automotive/ccp.uts#L959-L992

File: scapy/contrib/automotive/can_xcp.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        self.__resource = resource
+        self.__comm_mode_basic = comm_mode_basic
+        self.__max_cto = max_cto
+        self.__max_dto = max_dto
+        self.__xcp_protocol_version = xcp_protocol_version
+        self.__xcp_transport_version = xcp_transport_version
+        self.__request_id = request_id
+        self.__response_id = response_id
+
+    def RESOURCE(self):
```
Comment (gpotter2) [inline scapy/contrib/automotive/can_xcp.py]: Shouldn't you use the `@property` decorator?

File: scapy/contrib/automotive/can_xcp.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+# scapy.contrib.status = loads
+
+import threading
+import time
+import scapy.modules.six as six
+from scapy.config import LINUX
+from scapy.layers.can import CAN
+
+
+class Verbose():
```
Comment (gpotter2) [inline scapy/contrib/automotive/can_xcp.py]: This looks unnecessary.
You could use directly the logging module, and use by default `log_runtime` from `Scapy.error`

File: scapy/contrib/automotive/can_xcp.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
+#! /usr/bin/env python
```
Comment (gpotter2) [inline scapy/contrib/automotive/can_xcp.py]: Please do not use shebangs when you don't need to.

**Comment by polybassa** [comment]
@Ita1pu What's the state of this PR?

**Comment by Ita1pu** [comment]
@polybassa Sorry I didn't have the time to do anything on this PR. But I will start with it the beginning of next week.

**Comment by Ita1pu** [comment]
I deleted the forke repo by accident. A new PR has been created --> #2469

**Pull request closed without merging**